src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override bool CanRead
+        {
+            get { return !_disposed; }
+        }
+
+        public override bool CanSeek
+        {
+            get { return false; }
+        }
+
+        public override bool CanWrite
+        {
+            get { return !_disposed; }
+        }
+
+        public override long Length
+        {
+            get { throw new NotImplementedException(); }
+        }
+
+        public override long Position
+        {
+            get { throw new NotImplementedException(); }
+            set { throw new NotImplementedException(); }
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                throw new ObjectDisposedException("ProducerConsumerStream");
+            }
+
+            while (count > 0)
+            {
+                if (_currentBuffer == WriteBufferWrapper.EmptyContainer)
+                {
+                    _currentBuffer = await _buffer.Task;
+                }
+
+                int bytesWritten = _currentBuffer.Write(buffer, offset, count);
+                count -= bytesWritten;
+                offset += bytesWritten;
+                if (_currentBuffer.Count == 0)
+                {
+                    _currentBuffer = WriteBufferWrapper.EmptyContainer;
+                }
+
+                _dataAvail.TrySetResult(bytesWritten);
+            }
+        }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException("buffer");
+            }
+            if (offset < 0 || offset > buffer.Length)
+            {
+                throw new ArgumentOutOfRangeException("offset");
+            }
+            if (count < 0 || count > buffer.Length - offset)
+            {
+                throw new ArgumentOutOfRangeException("count");
+            }
+            Contract.EndContractBlock();
+
+            if (_disposed)
+            {
+                return 0;
+            }
```

Comment:
The only way to signal that there is nothing more to write is by calling `Dispose()`. From the writers perspective, once it has written everything and the final `WriteAsync()` has completed, the stream has been completed and it's free to call `Dispose()`. From the reader's perspective, the way it is signaled that a steam has finished and there's nothing else left is when a call to `Read()`/`ReadAsync()` returns zero bytes. [Here](https://msdn.microsoft.com/en-us/library/system.io.stream.read(v=vs.110).aspx) in the msdn documentation for the return value for Stream.Read it says:
> The total number of bytes read into the buffer. This can be less than the number of bytes requested if that many bytes are not currently available, or zero (0) if the end of the stream has been reached.  

Think of the scenario where `ReadAsync()` is called with a buffer which can hold 1024 bytes. `WriteAsync()` is called with 512 bytes, which is then copied to the read buffer then returns. The writer then calls `Dispose()` to signal it's done, but there's still a pending `ReadAsync()`. Okay, we can trigger the read to complete with a partially filled buffer (512 bytes), but the reader has no way of knowing that there's no more data. It could have been that a timer expired and wrote what it had so far. So the reader will call `ReadAsync()` again. If this throws an `ObjectDisposedException`, that could abort whatever it is that the reader is trying to do. I could add an extra state variable that after `Dispose()` is called, it allows one more final `Read` call which returns 0 bytes, after which it will throw `ObjectDisposedException`. What do you think? As it stands, further calls to `Write()` will throw `ObjectDisposedException` and further calls to `Read` will return immediately with 0, stating we're at the end of the stream. As no real resources are actually held by the class, I feel it's okay to have this half-disposed nature.